File: Oryx.Cognite/src/Beta/Subscriptions.fs
// Copyright 2023 Cognite AS
// SPDX-License-Identifier: Apache-2.0

namespace Oryx.Cognite.Beta

open System.Collections.Generic

open Oryx
open Oryx.Cognite
open Oryx.Cognite.Beta

open CogniteSdk
open CogniteSdk.Beta

[<RequireQualifiedAccess>]
module Subscriptions =
    [<Literal>]
    let Url = "/timeseries/subscriptions"

    /// Create a list of datapoint subscriptions
    let create (items: SubscriptionCreate seq) (source: HttpHandler<unit>) : HttpHandler<Subscription seq> =
        source
        |> withLogMessage "timeseries:subscriptions:create"
        |> withBetaHeader
        |> HttpHandler.create items Url

    /// Delete list of datapoint subscriptions by externalId, optionally ignoring unknown IDs
    let delete (items: string seq) (ignoreUnknownIds: bool) (source: HttpHandler<unit>) : HttpHandler<EmptyResponse> =
        http {
            let items = items |> Seq.map (fun f -> CogniteExternalId(f))

            let request =
                ItemsWithIgnoreUnknownIds(Items = items, IgnoreUnknownIds = ignoreUnknownIds)

            return!
                source
                |> withLogMessage "timeseries:subscriptions:delete"
                |> withBetaHeader
                |> HttpHandler.delete request Url
        }

    /// Retrieve the next batch of data from a datapoint subscription
    let listData (query: ListSubscriptionData) (source: HttpHandler<unit>) : HttpHandler<SubscriptionDataResponse> =
        source
        |> withLogMessage "timeseries:subscriptions:listdata"
        |> withBetaHeader
        |> HttpHandler.postV10 query (Url +/ "data/list")

    /// List timeseries in a datapoint subscription with pagination
    let listMembers
        (query: ListSubscriptionMembers)
        (source: HttpHandler<unit>)
        : HttpHandler<ItemsWithCursor<WrappedTimeSeriesId>> =
        source
        |> withLogMessage "timeseries:subscriptions:listmembers"
        |> withBetaHeader
        |> HttpHandler.getWithQuery query (Url +/ "members")

    /// List subscriptions in a project with pagination
    let list (query: ListSubscriptions) (source: HttpHandler<unit>) : HttpHandler<ItemsWithCursor<Subscription>> =
        source
        |> withLogMessage "timeseries:subscriptions:list"
        |> withBetaHeader
        |> HttpHandler.getWithQuery query Url

    /// Retrieve subscriptions by externalId, optionally ignoring unknown IDs
    let retrieve
        (items: string seq)
        (ignoreUnknownIds: bool)
        (source: HttpHandler<unit>)
        : HttpHandler<Subscription seq> =
        http {
            let items = items |> Seq.map (fun f -> CogniteExternalId(f))

            let request =
                ItemsWithIgnoreUnknownIds(Items = items, IgnoreUnknownIds = ignoreUnknownIds)

            let! ret =
                source
                |> withLogMessage "timeseries:subscriptions:retrieve"
                |> withBetaHeader
                |> HttpHandler.postV10<_, ItemsWithoutCursor<_>> request (Url +/ "byids")

            return ret.Items
        }

    /// Update a list of subscriptions
    let update
        (query: IEnumerable<UpdateItem<SubscriptionUpdate>>)
        (source: HttpHandler<unit>)
        : HttpHandler<Subscription seq> =
        source
        |> withLogMessage "timeseries:subscriptions:update"
        |> withBetaHeader
        |> HttpHandler.update query Url

File: CogniteSdk/src/Resources/Beta/Subscriptions.cs
// Copyright 2023 Cognite AS
// SPDX-License-Identifier: Apache-2.0

using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using System.Threading;

using CogniteSdk.Beta;

using Oryx;
using Oryx.Cognite.Beta;

using Microsoft.FSharp.Core;

namespace CogniteSdk.Resources.Beta
{
    /// <summary>
    /// Resource for timeseries subscriptions
    /// </summary>
    public class SubscriptionsResource : Resource
    {
        /// <summary>
        /// Constructor
        /// </summary>
        public SubscriptionsResource(Func<CancellationToken, Task<string>> authHandler, FSharpFunc<IHttpNext<Unit>, Task<Unit>> ctx) : base(authHandler, ctx)
        {
        }

        /// <summary>
        /// Create a list of datapoint subscriptions.
        /// </summary>
        /// <param name="subscriptions">Subscriptions to create</param>
        /// <param name="token">Optional cancellation token</param>
        /// <returns>Created subscriptions</returns>
        public async Task<IEnumerable<Subscription>> CreateAsync(IEnumerable<SubscriptionCreate> subscriptions, CancellationToken token = default)
        {
            var req = Subscriptions.create(subscriptions, GetContext(token));
            return await RunAsync(req).ConfigureAwait(false);
        }

        /// <summary>
        /// Delete a list of datapoint subscriptions, optionally ignoring unknown IDs
        /// </summary>
        /// <param name="items">External IDs of subscriptions to delete</param>
        /// <param name="ignoreUnknownIds">True to ignore IDs not present in CDF,
        /// if this is false and subscriptions are missing, the request will fail.</param>
        /// <param name="token">Optional cancellation token</param>
        /// <returns></returns>
        public async Task DeleteAsync(IEnumerable<string> items, bool ignoreUnknownIds = false, CancellationToken token = default)
        {
            var req = Subscriptions.delete(items, ignoreUnknownIds, GetContext(token));
            await RunAsync(req).ConfigureAwait(false);
        }

        /// <summary>
        /// Retrieve the next batch of data from one or more subscription partitions.
        /// </summary>
        /// <param name="query">Data retrieval query</param>
        /// <param name="token">Optional cancellation token</param>
        /// <returns>Any changes to the subscription partition(s) in the request</returns>
        public async Task<SubscriptionDataResponse> ListDataAsync(ListSubscriptionData query, CancellationToken token = default)
        {
            var req = Subscriptions.listData(query, GetContext(token));
            return await RunAsync(req).ConfigureAwait(false);
        }

        /// <summary>
        /// List timeseries in the given subscription, with pagination.
        /// </summary>
        /// <param name="query">Query</param>
        /// <param name="token">Optional cancellation token</param>
        /// <returns>A list of timeseries IDs and a cursor for pagination if there are more results</returns>
        public async Task<ItemsWithCursor<WrappedTimeSeriesId>> ListMembersAsync(ListSubscriptionMembers query, CancellationToken token = default)
        {
            var req = Subscriptions.listMembers(query, GetContext(token));
            return await RunAsync(req).ConfigureAwait(false);
        }

        /// <summary>
        /// List subscriptions in the project, with pagination.
        /// </summary>
        /// <param name="query">Query</param>
        /// <param name="token">Optional cancellation token</param>
        /// <returns>A list of subscriptions, and a cursor for pagination if there are more results</returns>
        public async Task<ItemsWithCursor<Subscription>> ListAsync(ListSubscriptions query, CancellationToken token = default)
        {
            var req = Subscriptions.list(query, GetContext(token));
            return await RunAsync(req).ConfigureAwait(false);
        }

        /// <summary>
        /// Retrieve a list of subscriptions by their external ID.
        /// </summary>
        /// <param name="items">Subscriptions to retrieve</param>
        /// <param name="ignoreUnknownIds">True to ignore IDs not present in CDF,
        /// if this is false and subscriptions are missing, the request will fail.</param>
        /// <param name="token">Optional cancellation token</param>
        /// <returns>Retrieved subscriptions</returns>
        public async Task<IEnumerable<Subscription>> RetrieveAsync(IEnumerable<string> items, bool ignoreUnknownIds = false, CancellationToken token = default)
        {
            var req = Subscriptions.retrieve(items, ignoreUnknownIds, GetContext(token));
            return await RunAsync(req).ConfigureAwait(false);
        }

        /// <summary>
        /// Update a list of subscriptions.
        /// </summary>
        /// <param name="updates">Subscription updates</param>
        /// <param name="token">Optional cancellation token</param>
        /// <returns>Updated subscriptions</returns>
        public async Task<IEnumerable<Subscription>> UpdateAsync(IEnumerable<UpdateItem<SubscriptionUpdate>> updates, CancellationToken token = default)
        {
            var req = Subscriptions.update(updates, GetContext(token));
            return await RunAsync(req).ConfigureAwait(false);
        }
    }
}
